### PR TITLE
[Android] - Add ability to detect if location was mocked

### PR DESCRIPTION
**Pull request by GantMan**
Since API 18, Android locations have had the `isFromMockProvider()` function, to verify the validity of a provided location.  This was one of many methods one could verify location data, but as of Marshmallow, the other ways of detecting if "Mock Locations" is on in developer settings has been deprecated or defunct.

This means some devices can only detect location mocking by exposing the method on the location object.

This change provides that exposure.

### CODE REVIEW

**Comment by facebook-github-bot-9** [comment]
@GantMan updated the pull request.

**Comment by GantMan** [comment]
@christopherdro  - any feedback?

**Comment by facebook-github-bot-6** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/b1e6e56be1db576e574d9fb78ae1cf7387736c96..c699db2c6984dbbd6a6b8abcb3fdb49b0a28f430)

**Comment by christopherdro** [comment]
LGTM. Maybe a space between the start and end of the conditional.

/cc @mkonicek Thoughts?

**Comment by GantMan** [comment]
@mkonicek interested in any feedback.
@christopherdro - would you like me to add that space?

**Comment by facebook-github-bot-3** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/c699db2c6984dbbd6a6b8abcb3fdb49b0a28f430..6410a3c01475413763966673cadf6c27e2cadc24)

**Comment by christopherdro** [comment]
@GantMan Sounds good

**Comment by satya164** [comment]
cc @dmmiller

**Comment by facebook-github-bot-8** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/6410a3c01475413763966673cadf6c27e2cadc24..7a2eabeda6eb8bdfebd57f58917760d0f15cf8c2)

**Comment by facebook-github-bot-6** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/7a2eabeda6eb8bdfebd57f58917760d0f15cf8c2..61f60791cdac5cf71e9de5b30b543c718f1f2ca7)

**Comment by GantMan** [comment]
@christopherdro  - whitespace added.
Ready to go unless we have other updates needed.

**Comment by satya164** [comment]
Just a heads-up, since RN polyfills the geolocation API according to web, and web doesn't have  a `mock` property, Flow won't have definitions for it, and if someone tries to use this, Flow will throw error. So we need to figure something out so that Flow passes.

Also, if we're adding this, we should also add it to the docs.

**Comment by GantMan** [comment]
@satya164  -   good point.  I've integrated this into a project that required it, and I didn't see any errors expressed, but they were probably covered up?   Not sure where flow errors are reported.

Happy to add to docs, I'll do so once we identify our approach.

Since the property is Android only, I'm somewhat stuck on where we can place this information (that comes from Android position, only).

The only way I can think to get this out of the position, so we stay conformed with web, is by adding a function that all positions get handed to, which for Android (or insert OS here) can then examine the position and set some kind of platform properties?   Something like `Platform.androidSpecific.lastLocationMocked()`.   Not in love with this.

Thoughts?

**Comment by satya164** [comment]
@GantMan 

> and I didn't see any errors expressed, but they were probably covered up

do you use Flow in your project? if you do Flow should give the error `Property not found on object type'.

> Something like Platform.androidSpecific.lastLocationMocked()

we won't want to add module specific stuff to `Platform`.

The only solution I can think of is to ship the flow definitions with React Native rather than relying on the definitions shipped with Flow.

**Comment by GantMan** [comment]
Is there any way to simply extend the web definitions?   Sort of like extending a babelrc?   That might make short work of this.   Sorry for my ignorance on flow, here.

**Comment by satya164** [comment]
I don't think it's currently possible (not sure though). But the Geolocation API is pretty small, so we could just copy it in.

https://github.com/facebook/flow/blob/master/lib/bom.js#L336

**Comment by GantMan** [comment]
@satya164 - I've updated the PR to include the flow RN polyfill.

**Comment by facebook-github-bot-4** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/61f60791cdac5cf71e9de5b30b543c718f1f2ca7..ccd451fe292d185ebd274a843e24bc429fb38a2f)

File: flow/position.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+declare class Position {
```
Comment (satya164) [inline flow/position.js]: Can you make it `Position.js` please?

**Comment by satya164** [comment]
Thanks. Can you also add it to the docs?

I'll run flow locally later to see if everything is okay

**Comment by facebook-github-bot-8** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/ccd451fe292d185ebd274a843e24bc429fb38a2f..b3a28bbd2d3dba1c1c0fea12dda1110a9d83e581)

**Comment by GantMan** [comment]
@satya164 - updated docs, and fixed case.   If you want me to document differently, or in another location, just let me know.

**Comment by facebook-github-bot-5** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/b3a28bbd2d3dba1c1c0fea12dda1110a9d83e581..e88b53d1e6b2c3ebbc4c25f844720a5938894a0f)

**Comment by facebook-github-bot-0** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/e88b53d1e6b2c3ebbc4c25f844720a5938894a0f..b8418544b6f5e639979bedc716079766ceae3b51)

**Comment by facebook-github-bot-4** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/b8418544b6f5e639979bedc716079766ceae3b51..fbf8bd193f59165895d17ebe9191f959670f3bec)

**Comment by facebook-github-bot-5** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/b8418544b6f5e639979bedc716079766ceae3b51..fbf8bd193f59165895d17ebe9191f959670f3bec)

**Comment by satya164** [comment]
Hey sorry, I got a bit busy. Will ship it in 2-3 days after testing.

**Comment by GantMan** [comment]
all good @satya164  I know how it is :)  Thanks for your continued efforts and for keeping RN moving!

**Comment by facebook-github-bot-2** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/fbf8bd193f59165895d17ebe9191f959670f3bec..36c90b60956b34e7b9c4f3fda7f3f488ef92bea3)

**Comment by facebook-github-bot-1** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/36c90b60956b34e7b9c4f3fda7f3f488ef92bea3..ae9470dd44f8078a93f1bae40b2f3fb665beb725)

**Comment by GantMan** [comment]
@satya164  - ping :)

File: flow/Position.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+declare class Position {
```
Comment (mkonicek) [inline flow/Position.js]: Where is this used?

**Comment by mkonicek** [comment]
Thanks for the review @satya164!

File: flow/Position.js (hunk 2 above)
Comment (GantMan) [inline flow/Position.js]: This is a flow mock to add the web polyfill for the mocked boolean.

**Comment by facebook-github-bot-4** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/ae9470dd44f8078a93f1bae40b2f3fb665beb725..ce8d89a384c36c6a8c9626beca8f6035d1184ece)

**Comment by GantMan** [comment]
@satya164  - merge conflict resolve, please test and merge soon :)  So no more conflicts.

**Comment by satya164** [comment]
Sorry for the delay. Will merge this tonight.

**Comment by satya164** [comment]
Tested and Flow is working fine.

**Comment by satya164** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-4** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3858205) to review internal test results.

**Comment by facebook-github-bot-6** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by satya164** [comment]
cc @bestander any idea what's wrong?

**Comment by bestander** [comment]
```
patching file fbobjc/Libraries/FBReactKit/js/react-native-github/Libraries/Geolocation/Geolocation.js
Hunk #1 FAILED at 30
1 out of 2 hunks FAILED -- saving rejects to file fbobjc/Libraries/FBReactKit/js/react-native-github/Libraries/Geolocation/Geolocation.js.rej
patching file fbandroid/java/com/facebook/catalyst/js/react-native-github/Libraries/Geolocation/Geolocation.js
```

**Comment by bestander** [comment]
Could you rebase?

**Comment by facebook-github-bot-4** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/ce8d89a384c36c6a8c9626beca8f6035d1184ece..899dee25596b2be707590355d1378889326f36d1) - [changes since last import](https://github.com/facebook/react-native/pull/9390/files/ce8d89a384c36c6a8c9626beca8f6035d1184ece..899dee25596b2be707590355d1378889326f36d1)

**Comment by GantMan** [comment]
just merged master.  Let me know what I can do.  Happy to help with whatever is visible to me ;) Cheers!

**Comment by satya164** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-7** [comment]
I tried to merge this pull request into the Facebook internal repo but some checks failed. To unblock yourself please check the following: Does this pull request pass all open source tests on GitHub? If not please fix those. Does the code still apply cleanly on top of GitHub master? If not can please rebase. In all other cases this means some internal test failed, for example a part of a fb app won't work with this pull request. I've added the Import Failed label to this pull request so it is easy for someone at fb to find the pull request and check what failed. If you don't see anyone comment in a few days feel free to comment mentioning one of the core contributors to the project so they get a notification.

**Comment by bestander** [comment]
I'll try it manually

**Comment by facebook-github-bot-4** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/899dee25596b2be707590355d1378889326f36d1..1df0baa6b29cb1df6f7df319be5de4c9c780e9b5) - [changes since last import](https://github.com/facebook/react-native/pull/9390/files/899dee25596b2be707590355d1378889326f36d1..1df0baa6b29cb1df6f7df319be5de4c9c780e9b5)

**Comment by facebook-github-bot-9** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/899dee25596b2be707590355d1378889326f36d1..1df0baa6b29cb1df6f7df319be5de4c9c780e9b5) - [changes since last import](https://github.com/facebook/react-native/pull/9390/files/899dee25596b2be707590355d1378889326f36d1..1df0baa6b29cb1df6f7df319be5de4c9c780e9b5)

**Comment by facebook-github-bot-8** [comment]
@GantMan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9390/files/1df0baa6b29cb1df6f7df319be5de4c9c780e9b5..2696f5d9cb9e23a4a11e5fb7e67adbe3d9322b82) - [changes since last import](https://github.com/facebook/react-native/pull/9390/files/899dee25596b2be707590355d1378889326f36d1..2696f5d9cb9e23a4a11e5fb7e67adbe3d9322b82)

**Comment by satya164** [comment]
Hey @bestander , any luck with manual merge?

**Comment by bestander** [comment]
Investigating, sent a message to OSS team.
There seams to be a problem with rebasing improted Diffs.

**Comment by GantMan** [comment]
Thanks for staying on top of this @bestander  and @satya164 !  ❤️ 

I have a project where mocked locations has become a real problem, and this fix is key.  I'm building from source at the moment, to bake this change in, but we all know how tricky and slow that can be.

As an alternative, I can always open a second PR with the needed code change, so no rebase is needed.

**Comment by bestander** [comment]
@facebook-github-bot shipit

**Comment by bestander** [comment]
The previous attempt failed with this rejection
```
--- Geolocation.js
+++ Geolocation.js
@@ -31,9 +31,11 @@
 }
 
 /**
- * The Geolocation API follows the web spec:
+ * The Geolocation API extends the web spec:
  * https://developer.mozilla.org/en-US/docs/Web/API/Geolocation
  *
+ * Additions are made where applicable.
+ *
  * ### iOS
  * You need to include the `NSLocationWhenInUseUsageDescription` key
  * in Info.plist to enable geolocation. Geolocation is enabled by default

```

**Comment by facebook-github-bot-7** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3858205) to review internal test results.

**Comment by GantMan** [comment]
awesome!   Thanks @bestander

**Pull request closed without merging**